signer/signer.go

```diff
@@ -0,0 +1,489 @@
+	switch req.PublicKey.(type) {
+	case *rsa.PublicKey:
+		if !p.allowRSAKeys {
+			return errors.New("RSA keys not allowed")
+		}
+	case *ecdsa.PublicKey:
+		if !p.allowECDSAKeys {
+			return errors.New("ECDSA keys not allowed")
+		}
+	default:
+		return errors.New("unsupported public key type")
+	}
+
+	if !p.allowMustStaple && req.IncludeMustStaple {
+		return errors.New("must-staple extension cannot be included")
+	}
+
+	if !p.allowCTPoison && req.IncludeCTPoison {
+		return errors.New("ct poison extension cannot be included")
+	}
+
+	if !p.allowSCTList && req.SCTList != nil {
+		return errors.New("sct list extension cannot be included")
+	}
+
+	if req.IncludeCTPoison && req.SCTList != nil {
+		return errors.New("cannot include both ct poison and sct list extensions")
+	}
+
+	if !p.allowCommonName && req.CommonName != "" {
+		return errors.New("common name cannot be included")
+	}
+
+	validity := req.NotAfter.Sub(req.NotBefore)
+	if validity <= 0 {
+		return errors.New("NotAfter must be after NotBefore")
+	}
+	if validity > p.maxValidity {
+		return fmt.Errorf("validity period is more than the maximum allowed period (%s>%s)", validity, p.maxValidity)
+	}
+	backdatedBy := clk.Now().Sub(req.NotBefore)
+	if backdatedBy > p.maxBackdate {
+		return fmt.Errorf("NotBefore is backdated more than the maximum allowed period (%s>%s)", backdatedBy, p.maxBackdate)
+	}
+	if backdatedBy < 0 {
+		return errors.New("NotBefore is in the future")
+	}
+
+	return nil
+}
+
+var defaultEKU = []x509.ExtKeyUsage{
+	x509.ExtKeyUsageServerAuth,
+	x509.ExtKeyUsageClientAuth,
+}
+
+func (p *signingProfile) generateTemplate(clk clock.Clock) *x509.Certificate {
+	template := &x509.Certificate{
+		SignatureAlgorithm:    p.sigAlg,
+		ExtKeyUsage:           defaultEKU,
+		OCSPServer:            []string{p.ocspURL},
+		IssuingCertificateURL: []string{p.issuerURL},
+		BasicConstraintsValid: true,
+	}
+
+	if p.crlURL != "" {
+		template.CRLDistributionPoints = []string{p.crlURL}
+	}
+
+	if p.policies != nil {
+		template.ExtraExtensions = []pkix.Extension{*p.policies}
+	}
+
+	return template
+}
+
+// Signer is a certificate signer
+type Signer struct {
+	issuer  *x509.Certificate
+	signer  crypto.Signer
+	profile *signingProfile
+	clk     clock.Clock
+	lintKey crypto.Signer
+	lints   lint.Registry
+}
+
+// Config contains the information necessary to construct a Signer
+type Config struct {
+	Issuer       *x509.Certificate
+	Signer       crypto.Signer
+	IgnoredLints []string
+	Clk          clock.Clock
+	Profile      ProfileConfig
+}
+
+// NewSigner constructs a Signer from the provided Config
+func NewSigner(config Config) (*Signer, error) {
+	profile, err := newProfile(config.Profile)
+	if err != nil {
+		return nil, err
+	}
+	lints, err := lint.GlobalRegistry().Filter(lint.FilterOptions{
+		ExcludeNames:   config.IgnoredLints,
+		ExcludeSources: []lint.LintSource{lint.EtsiEsi}, // we don't care about the ETSI lints
```

Comment:
This seems like a new change relative to cfssl; is that right? Were the ETSI lints causing problems?

Reply:
They are not inexpensive, and don't apply to us, including them is just a waste of cycles.

Reply:
Maybe also worth excluding `lint.CABFEVGuidelines` too?